### ISSUE TITLE
Select All can cause immediate abnormal termination

### ISSUE THREAD

**Issue by pmor13** [mintty/mintty#1134]
Select All can cause immediate abnormal termination.
Possibly related to memory allocation failure.
mintty 3.5.1 (x86_64_pc-cygwin).

**Comment by mintty** [comment]
Not for me. More details are needed to reproduce an issue.

**Comment by PavelMorozkin** [comment]
What will happen here if renewn returns NULL?
```
// clip_addchar
b->text = renewn(b->text, b->capacity);
```

**Comment by mintty** [comment]
Admittedly, the case of exhausted memory isn't handled everywhere in mintty.
Anyway, did you check that this is actually the cause of your case? Can I assume thus that your scrollback buffer is particularly large (and your memory small for some reason)? Better diagnostic information leads to better and/or faster problem fixes...

**Comment by mintty** [comment]
... waiting for a reproducible test case

**Comment by PavelMorozkin** [comment]
I don't have (now) a reproducible test case.
I am not sure (now) that this is actually the cause of my case.
I think that if `renewn` returns `NULL`, then:
1. An appropriate diagnostics needs to be produced.
2. The operation needs to be cancelled.

**Comment by mintty** [comment]
You could e.g. upload your mintty config file; of particular interest is the setting of ScrollbackLines. Also you could answer my question about your system situation. Particularly many programs running? Low memory equipped?

Of course I agree in theory that checking of memory allocation failure should be consistently stringent and I had considered a code revision on this before but there are a bit too many of those to make it a pleasant endeavour...

**Comment by PavelMorozkin** [comment]
mintty config file (`.minttyrc:`):
```
FontHeight=12
Columns=159
Font=Consolas
FontSmoothing=default
FontWeight=400
FontIsBold=no
ForegroundColour=238,238,238
CopyTab=yes
ScrollbackLines=9999999
```
> many programs running?
No.

> Low memory equipped?
No.

Before "Select All" there was a long session. However, not sure that `ScrollbackLines` was reached / exceeded.

**Comment by mintty** [comment]
Thanks;

> immediate abnormal termination

Also please describe what you mean by "immediate" and by "abnormal termination". Details might be significant for further diagnostics. I have tested selection of very large scrollback to grow over-linearly in duration, but no termination so far.

**Comment by mintty** [comment]
This is in fact a case of virtual memory exhaustion. I'm hitting it when the internal memory need for clipboard preparation exceeds 24GB. After acquiring the clipboard contents, additional memory will be consumed for generating clipboard formats (UTF-8 text, RTF and HTML if enabled), so simply checking for out of memory at the point you pointed out does not solve the issue. I'm therefore considering to tame the issue by limiting the scrollback buffer size, e.g. to 99999 if that's suffiicient; also based by the observation that copying starts to get significantly slower from a few 100000 lines.

**Comment by mintty** [comment]
Released 3.5.3.

**Comment by pmor13** [comment]
> what you mean by "immediate" and by "abnormal termination":

After "Select All" was pressed (selected) the mintty terminated immediately w/o any diagnostic message (hence, the "abnormal termination").
